oeps/oep-0038-Data-Modeling.rst

```diff
@@ -0,0 +1,411 @@
+
+-  History using django-simple-history_. .. _django-simple-history: https://django-simple-history.readthedocs.io/en/latest/
+
+-  Remember to `backfill history`_. for existing models.
+
+.. _backfill history: https://django-simple-history.readthedocs.io/en/latest/quick_start.html#existing-projects
+
+-  Where Django simple history is not an option, the following data should be captured:
+
+   -  Fields that were changed
+
+   -  Date & time of the change
+
+   -  The foreign key of the user who initiated the change
+
+Use the correct data type for a column
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  Don’t use a :code:`IntegerField` when a :code:`BooleanField` would do.
+
+-  Use :code:`BigIntegerField`_. for your foreign keys
+
+-  Don’t store an Integer field as :code:`CharField`_..
+.. _CharField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#charfield
+
+- Store UUID's as :code:`UUIDField`_. with a max length that matches the max length of the UUID.
+.. _UUIDField: https://docs.djangoproject.com/en/2.2/ref/models/fields/#uuidfield
+
+
+Each column in a table should only store a single fact or dimension
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- If a column could be a mix of integer data and character data it is best to store these items as 2 different fields in the database
+
+Models should be Annotated for PII
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+-  All models in the Open edX ecosystem should be tagged for PII using `code annotations`_. by following OEP-30_.
+.. _code annotations: https://github.com/edx/code-annotations>
+.. _OEP-30: https://github.com/edx/open-edx-proposals/blob/master/oeps/oep-0030-arch-pii-markup-and-auditing.rst
+
+
+Have sane default values
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+- A model should have default values that make sense for the application
+
+- :code:`CharField`_. Should be defined with the :code:`null=True`_. option.
+.. _null=True: https://docs.djangoproject.com/en/2.2/ref/models/fields/#null
+   - This is to improve post processing. Defaulting :code:`CharField`_. to null enables us to better derive the intent of the user. If the field is null no intent was made to enter that field. If the field is blank a string was entered and was modified later by the user.
+
+- For example if you are adding a boolean to flag that a learner has not yet activated their account, the default value should be set to False, not None.
+
+The database layer should preserve uniqueness
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- If a model needs to preserve uniqueness between many fields use :code:`unique_together`_.
+.. _unique_together: https://docs.djangoproject.com/en/2.2/ref/models/options/#unique-together
+
+
+
+Best Practices
+--------------
+These practices are designed to help teams create rich and efficient data models within the OpenedX ecosystem.
+They are not standards but guidelines to help teams think about how they are storing data.
```

Comment:
Should there be a best practice added about *not* using SQL/Snowflake keywords as column names in this section?